ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/electra/statetransition/epoch/EpochProcessorElectra.java

```diff
@@ -186,73 +201,176 @@ public void processEffectiveBalanceUpdates(
+      final BLSPublicKey pubkey, final Bytes32 withdrawalCredentials, final UInt64 amount) {
+    final Validator validator =
+        new Validator(
+            pubkey,
+            withdrawalCredentials,
+            ZERO,
+            false,
+            FAR_FUTURE_EPOCH,
+            FAR_FUTURE_EPOCH,
+            FAR_FUTURE_EPOCH,
+            FAR_FUTURE_EPOCH);
+
+    final UInt64 maxEffectiveBalance =
+        stateAccessorsElectra.getValidatorMaxEffectiveBalance(validator);
+    final UInt64 validatorEffectiveBalance =
+        amount
+            .minusMinZero(amount.mod(specConfig.getEffectiveBalanceIncrement()))
+            .min(maxEffectiveBalance);
+
+    return validator.withEffectiveBalance(validatorEffectiveBalance);
+  }
+
+  /** process_pending_deposits */
+  @Override
+  public void processPendingDeposits(final MutableBeaconState state) {
+    final MutableBeaconStateElectra stateElectra = MutableBeaconStateElectra.required(state);
+
+    final UInt64 nextEpoch = beaconStateAccessors.getCurrentEpoch(state).plus(UInt64.ONE);
     final UInt64 availableForProcessing =
         stateElectra
             .getDepositBalanceToConsume()
             .plus(stateAccessorsElectra.getActivationExitChurnLimit(stateElectra));
-
     UInt64 processedAmount = UInt64.ZERO;
     int nextDepositIndex = 0;
-    final List<PendingBalanceDeposit> depositsToPostpone = new ArrayList<>();
+    final List<PendingDeposit> depositsToPostpone = new ArrayList<>();
+    boolean isChurnLimitReached = false;
+    final UInt64 finalizedSlot =
+        miscHelpers.computeStartSlotAtEpoch(stateElectra.getFinalizedCheckpoint().getEpoch());
+
+    for (final PendingDeposit deposit : stateElectra.getPendingDeposits()) {
+      // Do not process deposit requests if Eth1 bridge deposits are not yet applied.
+      final boolean isDepositRequest = deposit.getSlot().isGreaterThan(GENESIS_SLOT);
+      final boolean hasPendingEth1BridgeDeposits =
+          stateElectra
+              .getEth1DepositIndex()
+              .isLessThan(stateElectra.getDepositRequestsStartIndex());
+      if (isDepositRequest && hasPendingEth1BridgeDeposits) {
+        break;
+      }
+
+      // Check if deposit has been finalized, otherwise stop processing
+      if (deposit.getSlot().isGreaterThan(finalizedSlot)) {
+        break;
+      }
 
-    final SszList<PendingBalanceDeposit> pendingBalanceDeposits =
-        stateElectra.getPendingBalanceDeposits();
-    for (final PendingBalanceDeposit deposit : pendingBalanceDeposits) {
-      final Validator validator = state.getValidators().get(deposit.getIndex());
+      // Check if number of processed deposits has not reached the limit, otherwise, stop processing
+      if (nextDepositIndex
+          >= SpecConfigElectra.required(specConfig).getMaxPendingDepositsPerEpoch()) {
+        break;
+      }
+
+      // TODO-lucas What is the impact of doing this? Should we use the transition cache?
```

Comment:
Hmm, yeah this for sure needs to change. Think we should use `ValidatorsUtil.getValidatorIndex()` (internally uses the cache) and then just use .get(index) on the validators in the state.

Reply:
Done!